### PR TITLE
GH Actions: Build platform-independent wheels of sagemath-environment, sage-setup, sage-sws2rst for PyPI

### PR DESCRIPTION
**Pull request by mkoeppe**
We already build platform wheels for **sagemath-{objects,categories,repl}** for PyPI using cibuildwheel.
Here we add building of platform-independent wheels for **sagemath-environment** and **sage-setup** and **sage-sws2rst** to the GH Actions workflow. (They are needed, for example, for pyodide / jupyterlite.)

To test locally: `make pypi-noarch-wheels`

We also update actions/upload-artifact, actions/download-artifact to v4. This requires a restructuring, as we can no longer upload the wheels for different architectures (built by separate matrix jobs) to the same artifact: https://github.com/actions/upload-artifact?tab=readme-ov-file#not-uploading-to-the-same-artifact
Instead we upload them as separate artifacts and deploy them to PyPI directly from the job that built them.

This is:
- part of #31251

### :memo: Checklist

- [x] The title is concise, informative, and self-explanatory.
- [x] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by mkoeppe** [comment]
@kiwifb Any thoughts on this one here?

**Comment by kiwifb** [comment]
I am not very hot on the subject of producing that kind of wheels. I am not qualified to review this. I should have said two weeks ago but I was sick.

**Comment by mkoeppe** [comment]
No worries!

File: .github/workflows/dist.yml [hunk 1]
```diff
@@ -103,6 +103,36 @@ jobs:
...
+        run: |
+          sudo DEBIAN_FRONTEND=noninteractive apt-get update
+          sudo DEBIAN_FRONTEND=noninteractive apt-get install $(build/bin/sage-get-system-packages debian _bootstrap)
+      - name: make pypi-noarch-wheels
+        run: |
+          ./bootstrap
+          ./configure
+          make pypi-noarch-wheels V=0
+          (mkdir dist && mv upstream/sage*-none-any.whl dist/)
+          ls -l dist
```
Comment (kwankyu) [inline .github/workflows/dist.yml:123]: Is this right?

I get 
```
Built wheels are in venv/var/lib/sage/wheels/
```

Comment (mkoeppe) [reply .github/workflows/dist.yml:123]: You are right.

Comment (mkoeppe) [reply .github/workflows/dist.yml:123]: I'll also update the actions while at it, hold on please...

**Comment by kwankyu** [comment]
Will you test the updated workflow somewhere?

**Comment by mkoeppe** [comment]
Running at https://github.com/mkoeppe/sage/actions/runs/8843181051/job/24282993945

**Comment by mkoeppe** [comment]
Completed successfully

**Comment by kwankyu** [comment]
The artifacts are now split according to architecures by the changed name `${{ matrix.os }}-${{ matrix.arch }}-wheels`.

I don't know where the artifacts are used downstream. Are they used? If so, this change may break there. No?

**Comment by mkoeppe** [comment]
The artifacts are not used by anything. They are uploaded only so that one can download them manually if something needs to be checked.

**Review by kwankyu — APPROVED** [review]
LGTM.

**Comment by mkoeppe** [comment]
Thank you!

**Pull request merged**